v4/newrelic/segments.go

```diff
@@ -126,14 +143,46 @@ const (
 	MessageExchange MessageDestinationType = "Exchange"
 )
 
+func (s *span) end() {
+	s.Span.End()
+	s.Lock()
+	s.ended = true
+	s.Unlock()
+	if s.txn != nil {
```

Comment:
Should we avoid doing this work when the segment was ended before?

I'm not sure what guarantees the Go agent makes for multiple calls to `segment.End`, but when multiple calls to `segment.End` are not caught, parenting might be off. For example in this case:
```go
txn := app.StartTransaction("transaction")
seg1 := txn.StartSegment("seg1")
seg2 := txn.StartSegment("seg2")
seg2.End()
seg1.End()
seg3 := txn.StartSegment("seg3")
seg2.End() // That's a weird one.
seg4 := txn.StartSegment("seg4")
seg4.End()
seg3.End()
txn.End()
```

I'm not sure to me whether the Go agent suggests that subsequent calls to `segment.End` have no effect. If not, feel free to ignore this comment!